### PR TITLE
chore: refactor CoValueState to manage the availability logic

### PR DESCRIPTION
**Pull request by gdorsi**
To gain a deeper understanding of the loading flow I've started a refactoring on `CoValueState` trying to centralize all the logic around the availability state management of the CoValues.

### CODE REVIEW

File: packages/cojson/src/localNode.ts [hunk 1]
```diff
@@ -302,10 +300,21 @@ export class LocalNode {
...
+
+        const updatedEntry = this.coValues[id];
+
+        if (updatedEntry?.state.type === "available") {
+            return Promise.resolve(updatedEntry.state.coValue);
+        }
+
+        delete this.coValues[id];
+
+        return "unavailable";
```
Comment (gdorsi) [inline packages/cojson/src/localNode.ts:315]: Not sure on what we should do when a CoValue is not available.

It could be the case of a CoValue that have not synced yet by another user but we've got already an invitation.

I think we should store all the unavailable coValues somewhere and periodically retry to load them (e.g. every 30s)

File: packages/cojson/src/sync.ts [hunk 2]
```diff
@@ -131,53 +132,19 @@ export class SyncManager {
...
 
-            const coValueEntry = this.local.coValues[id];
-            if (coValueEntry?.state !== "loading") {
-                continue;
-            }
-            const firstStateEntry = coValueEntry.firstPeerState[peer.id];
-            if (firstStateEntry?.type !== "waiting") {
-                throw new Error("Expected firstPeerState to be waiting " + id);
+            if (coValueEntry?.state.type === "unknown") {
+                await coValueEntry.state.waitForPeer(peer.id);
```
Comment (gdorsi) [inline packages/cojson/src/sync.ts:146]: Most of the logic that was here has been moved inside CoValueState

File: packages/cojson/src/sync.ts [hunk 3]
```diff
@@ -514,39 +478,25 @@ export class SyncManager {
...
-                )
-            ) {
-                entry.resolve("unavailable");
+
+            if (!availableOnPeer) {
+                entry.dispatch({
+                    type: "not-found",
+                    peerId: peer.id,
+                });
             }
```
Comment (gdorsi) [inline packages/cojson/src/sync.ts:489]: I'm not updating the peer state to "available" because:
- notifying the listeners by resolving the peer state promise breaks the sync
- the coValue is thruly available in the peer if we have got it's headers

File: packages/cojson/src/sync.ts [hunk 4]
```diff
@@ -566,35 +514,27 @@ export class SyncManager {
...
-                onProgress: entry.onProgress,
-            };
+            coValue = new CoValueCore(msg.header, this.local);
 
-            this.local.coValues[msg.id] = entry;
+            entry.dispatch({
+                type: "found",
+                coValue,
+                peerId: peer.id,
+            });
```
Comment (gdorsi) [inline packages/cojson/src/sync.ts:533]: Most of the logic here have been moved inside the dispatch

File: packages/cojson/src/sync.ts [hunk 5]
```diff
@@ -648,15 +588,13 @@ export class SyncManager {
...
-            );
-
-            entry.onProgress?.(ourTotalnTxs / theirTotalnTxs);
+            // const theirTotalnTxs = Object.values(
+            //     peer.optimisticKnownStates[msg.id]?.sessions || {},
+            // ).reduce((sum, nTxs) => sum + nTxs, 0);
+            // const ourTotalnTxs = [...coValue.sessionLogs.values()].reduce(
+            //     (sum, session) => sum + session.transactions.length,
+            //     0,
+            // );
```
Comment (gdorsi) [inline packages/cojson/src/sync.ts:597]: Removed the `onProgress` hook because it was unused. (happy to restore it if I'm wrong)

Keeping the code commented because we could reuse it to improve the load flow and resolve only when all the known sessions have been loaded.

File: packages/cojson/src/sync.ts [hunk 6]
```diff
@@ -680,10 +618,6 @@ export class SyncManager {
             );
         }
 
-        if (resolveAfterDone) {
-            resolveAfterDone(coValue);
-        }
```
Comment (gdorsi) [inline packages/cojson/src/sync.ts:685]: Moved the resolve up.

Promises are resolved on the next tick of the event loop anyway so, since there are no await in the middle, it shouldn't matter where we resolve the covalue state.

File: packages/cojson/src/coValueState.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+    return { promise, resolve };
+}
+
+class CoValueUnknownState {
+    type = "unknown" as const;
+    private peers: Map<PeerID, ReturnType<typeof createResolvablePromise<"available" | "unavailable">>>;
+    private resolve: (value: "available" | "unavailable") => void;
+
+    ready: Promise<"available" | "unavailable">;
```
Comment (gdorsi) [inline packages/cojson/src/coValueState.ts:19]: Renamed the states:
 'loading' -> 'unknown'
 'loaded' -> 'available'
 
Reasoning about the names helped me in the understanding process.
Happy to revert the change if the previous ones have a better fit in this context.

Comment (aeplay) [reply packages/cojson/src/coValueState.ts:19]: This is better, what was perviously "loaded" isn't actually loaded, it's just "we've received the first content for it" - so "available" is much better

In the future we should have some kind of "loaded" state meaning "up-to-date as per all peer states"

File: packages/cojson/src/localNode.ts (hunk 1 above)
Comment (aeplay) [reply packages/cojson/src/localNode.ts:315]: I'd say exponential retry starting with 1s up to a max period of 30s - please create another ticket for that

File: packages/cojson/src/sync.ts (hunk 2 above)
Comment (aeplay) [reply packages/cojson/src/sync.ts:146]: Wonderful, this helps untangling things

**Review by aeplay — APPROVED** [review]
Great improvement!

File: packages/cojson/src/localNode.ts (hunk 1 above)
Comment (gdorsi) [reply packages/cojson/src/localNode.ts:315]: https://github.com/gardencmp/jazz/issues/598
